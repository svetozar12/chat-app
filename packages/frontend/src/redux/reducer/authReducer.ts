import { Action } from "../state";
import { ActionType } from "../state";
import { InitialState } from "../state";
import { HYDRATE } from "next-redux-wrapper";

const initialState = {
  remember_me: false,
  input: "",
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: "",
};
const reducer = (state: InitialState = initialState, action: Action) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.SAVE_INPUT:
      return { ...state, input: action.payload };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    case ActionType.SIGN_OUT:
      return { state: undefined };
    default:
      return state;
  }
};

export default reducer;
