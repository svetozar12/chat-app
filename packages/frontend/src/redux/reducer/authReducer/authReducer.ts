import { HYDRATE } from "next-redux-wrapper";
import { IAuthState } from "./state";
import { ActionType } from "../../types";
export const initialState: IAuthState = {
  remember_me: false,
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: {
    username: "",
    id: "",
    token: "",
    refresh_token: "",
  },
};
const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default reducer;
export { initialState as AuthState };
