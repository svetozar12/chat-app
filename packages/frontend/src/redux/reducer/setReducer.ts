import { Action, ActionType, InitialState2 } from "../state";
export const initialState = {
  reciever: "",
  chat_inviter: "",
  pageNumber: 2,
  setUserSettings: false,
  setFriendRequest: false,
  toggleCreateGroup: false,
  setMobileNav: false,
  setIsMatch: false,
  setChatSettings: false,
  setModalInvite: false,
  setIsLoggingIn: false,
};
const setReducer = (state: InitialState2 | any | InitialState2[] = initialState, action: Action) => {
  switch (action.type) {
    case ActionType.SET_RECIEVER:
      return { ...state, reciever: action.payload };
    case ActionType.CHAT_INVITER:
      return { ...state, chat_inviter: action.payload };
    case ActionType.INCREMENT_PAGE_NUMBER:
      return { ...state, pageNumber: ++action.payload };
    case ActionType.SET_FRIEND_REQUEST:
      return { ...state, setFriendRequest: action.payload };
    case ActionType.SET_USER_SETTINGS:
      return { ...state, setUserSettings: action.payload };
    case ActionType.TOGGLE_CREATE_GROUP:
      return { ...state, toggleCreateGroup: action.payload };
    case ActionType.SET_MOBILE_NAV:
      return { ...state, setMobileNav: action.payload };
    case ActionType.SET_CHAT_SETTINGS:
      return { ...state, setChatSettings: action.payload };
    case ActionType.SET_IS_MATCH:
      return { ...state, setIsMatch: action.payload };
    case ActionType.SET_MODAL_INVITE:
      return { ...state, setModalInvite: action.payload };
    case ActionType.SET_IS_LOGEDIN:
      return { ...state, setIsLoggingIn: action.payload };
    case ActionType.SIGN_OUT:
      return initialState;
    default:
      return state;
  }
};

export default setReducer;
