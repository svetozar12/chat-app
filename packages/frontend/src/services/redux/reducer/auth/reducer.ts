import { HYDRATE } from 'next-redux-wrapper';
import { IAuth } from './state';
import { ActionType } from '../../types';

export const initialState: IAuth = {
  remember_me: false,
  loginPrompt: false,
};
const authReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default authReducer;
