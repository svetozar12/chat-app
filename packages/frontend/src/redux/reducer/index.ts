import { combineReducers } from "redux";
import reducer from "./authReducer";
import setReducer from "./setReducer";
import saveInputReducer from "./save_inputReducer";
import messageReducer from "./messageReducer";

const reducers = combineReducers({
  authReducer: reducer,
  setReducer,
  messageReducer,
  saveInputReducer,
});

export default reducers;
