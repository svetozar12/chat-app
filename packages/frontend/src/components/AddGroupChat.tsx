import React from "react";
import axios from "axios";
import { Socket } from "socket.io-client";
import { InitialState2 } from "../redux/state";
import { useSelector, useDispatch } from "react-redux";
import styled from "@emotion/styled";
const Search_bar = styled.input`
  width: 70%;
  height: 2rem;
  margin: 0.5rem 0;
  border: 1px solid var(--input-border-color);
  border-radius: 5px;
  transition: 0.3s;
  padding: 1.3rem 0.9rem;
`;

const Button = styled.button`
  border-radius: 5px;
  border: none;
  padding: 0.8rem;
  cursor: pointer;
  transition: 0.4s;
`;
const Form = styled.form`
  display: flex;
  flex-direction: row;
`;
function AddGroupChat({
  cookieName,
  socketRef,
}: {
  cookieName: string;
  socketRef: Socket;
}) {
  const [user, setUser] = React.useState<string>("");
  const [usersData, setUsersData] = React.useState<string[]>([]);

  const dispatch = useDispatch();

  const state1 = useSelector(
    (state: { setReducer: InitialState2 }) => state.setReducer,
  );

  const emitFriendRequest = async () => {
    socketRef?.emit("friend_request");
  };

  const addToGroup = (user: string, e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setUsersData((prev) => [...prev, user]);
    setUser("");
  };
  const handleSumbit = async () => {
    try {
      const result = usersData.includes(cookieName);

      if (!result) usersData.push(cookieName);
      await axios.post("http://localhost:4002/invites/group-chat", {
        usersData,
      });
      emitFriendRequest();
      setUsersData([]);
      dispatch({
        type: "TOGGLE_CREATE_GROUP",
        payload: !state1.toggleCreateGroup,
      });
      return true;
    } catch (error) {
      setUsersData([]);
      return false;
    }
  };

  return (
    <>
      {state1.toggleCreateGroup && (
        <div>
          <Form onSubmit={(e) => addToGroup(user, e)}>
            <Search_bar
              onChange={(e) => setUser(e.target.value)}
              value={user}
              placeholder="Add user ..."
              type="search"
            />
            <Button onClick={handleSumbit} type="button">
              Create room
            </Button>
          </Form>
          <div style={{ display: "flex" }}>
            {usersData.map((element, index) => {
              return (
                <p style={{ color: "black" }} key={index}>
                  {element}
                </p>
              );
            })}
          </div>
        </div>
      )}
    </>
  );
}

export default AddGroupChat;
