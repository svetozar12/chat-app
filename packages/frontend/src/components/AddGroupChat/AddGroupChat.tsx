import React from 'react';
import { css, cx } from '@emotion/css';
import { CloseButton, HStack, IconButton, Input, Spacer } from '@chakra-ui/react';
import { gqlSdk } from '@chat-app/sdk';
import { useCookie } from 'next-cookie';
import { getAuth } from 'utils/authMethods';
import useThemeColors from 'hooks/useThemeColors';
import { IoCreateOutline } from 'react-icons/io5';
import { connect } from 'react-redux';
import { STATE } from 'services/redux/reducer';
import { IWebSocket } from 'services/redux/reducer/websocket/state';
import { IToggle } from 'services/redux/reducer/toggles/state';
import { bindActionCreators, Dispatch } from 'redux';
import { toggleCreateGroup } from 'services/redux/reducer/toggles/actions';

interface IAddGroupChat {
  ws: IWebSocket;
  toggle: IToggle;
  toggleCreateGroup: typeof toggleCreateGroup;
}

function AddGroupChat(props: IAddGroupChat) {
  const { toggle, ws, toggleCreateGroup } = props;
  const [user, setUser] = React.useState<string>('');
  const [usersData, setUsersData] = React.useState<string[]>([]);
  const cookie = useCookie();
  const cookieName: string = cookie.get('name');

  const emitFriendRequest = async () => {
    ws.ws?.emit('friend_request');
  };

  const addToGroup = (user: string, e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setUsersData((prev) => [...prev, user]);
    setUser('');
  };

  const removeFromGroup = (username: string) => {
    const newUsersData = usersData.filter((user) => user !== username);
    setUsersData(newUsersData);
  };

  const handleSumbit = async () => {
    try {
      getAuth();
      const result = usersData.includes(cookieName);

      if (!result) usersData.push(cookieName);
      await gqlSdk.invite.createGroupChat(usersData);
      emitFriendRequest();
      setUsersData([]);
      toggleCreateGroup(!toggle.toggleCreateGroupModal);
      return true;
    } catch (error) {
      setUsersData([]);
      return false;
    }
  };

  const {
    colors: { chatMessageBgColor, chatBg },
  } = useThemeColors();

  return (
    <>
      {toggle.toggleCreateGroupModal && (
        <div
          className={css`
            width: 100%;
            height: 100%;
            display: flex;
            flex-direction: column;
            align-items: flex-start;
            justify-content: center;
            background: ${chatBg};
          `}
        >
          <form
            className={cx(
              css`
                width: 100%;
                flex-direction: row;
              `,
              'flex',
            )}
            onSubmit={(e) => addToGroup(user, e)}
          >
            <Input
              className={css`
                width: 70%;
                margin: 0.5rem 0;
                transition: 0.3s;
                padding: 1.3rem 0.9rem;
              `}
              borderTopRightRadius="none"
              borderBottomRightRadius="none"
              onChange={(e) => setUser(e.target.value)}
              value={user}
              placeholder="Add user ..."
              type="search"
            />
            <IconButton
              w="3rem"
              borderTopLeftRadius="none"
              borderBottomLeftRadius="none"
              aria-label="icon"
              icon={<IoCreateOutline />}
              onClick={handleSumbit}
            />
          </form>
          <HStack mx={5} gap={5}>
            {usersData.map((element, index) => (
              <HStack align="center" justify="center" bg={chatMessageBgColor} color="main_white" p="0.5rem 0.8rem" key={index}>
                <p>{element}</p>
                <Spacer />
                <CloseButton
                  onClick={() => {
                    removeFromGroup(element);
                  }}
                  borderRadius="full"
                />
              </HStack>
            ))}
          </HStack>
        </div>
      )}
    </>
  );
}

const mapStateToProps = (state: STATE) => ({
  ws: state.ws,
  toggle: state.toggle,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  toggleCreateGroup: bindActionCreators(toggleCreateGroup, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddGroupChat);
