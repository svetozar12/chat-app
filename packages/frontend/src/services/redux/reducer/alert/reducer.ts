import { AnyAction } from 'redux';
import { IAlert } from 'services/redux/reducer/alert/state';
import { ActionType } from '../../types';

const initialState: IAlert = {
  good: '', // good and bad stand for good alert and bad alert
  bad: '',
};

const alertReducer = (state = initialState, action: AnyAction) => {
  switch (action.type) {
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.payload };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.payload };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.payload };
    default:
      return state;
  }
};

export default alertReducer;
