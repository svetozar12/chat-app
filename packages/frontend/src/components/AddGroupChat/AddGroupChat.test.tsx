import configureStore from "redux-mock-store";
import AddGroupChat from "components/AddGroupChat/AddGroupChat";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { initialState } from "redux/reducer/setReducer";
import { AuthState } from "redux/reducer/authReducer";
import { screen, cleanup } from "@testing-library/react";
import { ReactTestRendererJSON } from "react-test-renderer";
import "@testing-library/jest-dom";

let component: ReactTestRendererJSON | ReactTestRendererJSON[] | null;
const socketRef: any = jest.fn();

beforeAll(() => {
  const mockStore = configureStore([]);
  const store = mockStore({
    authReducer: AuthState,
    setReducer: initialState,
  });

  component = renderer
    .create(
      <Provider store={store}>
        <AddGroupChat cookieName="ivan" socketRef={socketRef} />
      </Provider>,
    )
    .toJSON();
});

afterAll(cleanup);

describe("Render connected React-redux page", () => {
  it("should create snapshot for <AddGroupChat/>", () => {
    expect(component).toMatchSnapshot();
  });

  it("should render <AddGroupChat/>", () => {
    const renderedComponent = screen.getByTestId("form");
    screen.getByRole("");
    expect(renderedComponent).toBeInTheDocument();
  });
});
