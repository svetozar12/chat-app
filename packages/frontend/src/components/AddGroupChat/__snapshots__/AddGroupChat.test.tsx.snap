// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render connected React-redux page should create snapshot for <AddGroupChat/> 1`] = `
<div
  className="css-1h7sx70"
>
  <form
    className="css-o2j9ze"
    onSubmit={[Function]}
  >
    <input
      className="css-1q3nqrs"
      onChange={[Function]}
      placeholder="Add user ..."
      type="search"
      value=""
    />
    <button
      className="css-csscj8"
      onClick={[Function]}
      type="button"
    >
      Create room
    </button>
  </form>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  />
</div>
`;
